wcfsetup/install/files/lib/system/package/plugin/AbstractFileDeletePackageInstallationPlugin.class.php

```diff
@@ -81,30 +71,67 @@ protected function handleDelete(array $items)
-                    but to package '" . PackageCache::getInstance()->getPackage($filePackageID)->package . "'."
-                );
+            if (!isset($groupedFiles[$application])) {
+                $groupedFiles[$application] = [];
             }
+            $groupedFiles[$application][] = $file;
+        }
 
-            $filePath = $this->getFilePath($file, $application);
-            if (\file_exists($filePath)) {
-                \unlink($filePath);
+        $logFiles = [];
+        foreach ($groupedFiles as $application => $files) {
+            $conditions = new PreparedStatementConditionBuilder();
+            $conditions->add("{$this->getFilenameTableColumn()} IN (?)", [$files]);
+            $conditions->add('application = ?', [$application]);
+            $conditions->add('packageID = ?', [$this->installation->getPackageID()]);
+
+            $sql = "SELECT  packageID, application, {$this->getFilenameTableColumn()}
+                    FROM    {$this->getLogTableName()}
+                    {$conditions}";
+            $searchStatement = WCF::getDB()->prepare($sql);
+            $searchStatement->execute($conditions->getParameters());
+
+            while ($row = $searchStatement->fetchArray()) {
+                if (!isset($logFiles[$row['application']])) {
+                    $logFiles[$row['application']] = [];
+                }
+                $logFiles[$row['application']][$row[$this->getFilenameTableColumn()]] = $row['packageID'];
             }
+        }
 
-            $deleteStatement->execute([
-                $this->installation->getPackageID(),
-                $file,
-            ]);
+        foreach ($groupedFiles as $application => $files) {
+            foreach ($files as $file) {
+                $filePackageID = $logFiles[$application][$file] ?? null;
+                if ($filePackageID !== null && $filePackageID != $this->installation->getPackageID()) {
+                    throw new \UnexpectedValueException(
+                        "'{$file}' does not belong to package '{$this->installation->getPackage()->package}'
+                        but to package '" . PackageCache::getInstance()->getPackage($filePackageID)->package . "'."
+                    );
+                }
+
+                $filePath = $this->getFilePath($file, $application);
+                if (\file_exists($filePath)) {
+                    \unlink($filePath);
+                }
+            }
+        }
+
+        $sql = "DELETE FROM {$this->getLogTableName()}
+                WHERE       packageID = ?
+                        AND {$this->getFilenameTableColumn()} = ?
+                        AND application = ?";
+        $deleteStatement = WCF::getDB()->prepare($sql);
+
+        WCF::getDB()->beginTransaction();
+        foreach ($logFiles as $application => $files) {
+            foreach ($files as $file => $packageID) {
+                $deleteStatement->execute([
+                    $packageID,
```

Comment:
This makes me feel uneasy. While I know that `$packageID === $this->installation->getPackageID()` at this point it is not actually clear. I suggest to either change this to `$this->installation->getPackageID()` or to add an assertion verifying this.